### PR TITLE
Support extended resources and ephemeral-storage for scale-from-zero specified in MachineClass NodeTemplate

### LINKED ISSUE #132: Allow a way to specify extended resources for scale-from-zero scenario
**What would you like to be added**:
There should be a mechanism in our autoscaler so that the user could specify any extended resources his nodes have and the autoscaler becomes aware of it, so that during scaling out a node group from zero it could consider that.

**Why is this needed**:
It has been noticed that currently the autoscaler couldn't scale a node group from zero if the pod is requesting an extended resource. This is happening because the nodeTemplate which the autoscaler creates doesn't have the extended resources specified.
However it is able to scale from one , because there the autoscaler can form nodeTemplate from an existing node.
There are ways in [AWS](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/README.md#:~:text=From%20version%201.14,storage%3A%20100G) and [Azure](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/azure/README.md#resources) implementation of autoscaler to specify such resources.

### PR DESCRIPTION
**Pull request by elankath**
**What this PR does / why we need it**:

Right now in scale-from-zero cases, the autoscaler does not respect ephemeral-storage  and extended resource specified in the `MachineClass.NodeTemplate` . Only the standard `cpu`, `gpu` and `memory` are picked up. Neither is there any support for custom extended resource which are fully ignored presently.

**Which issue(s) this PR fixes**:
Fixes #132

**Special notes for your reviewer**:

**Release note**:

```feature operator
Support extended resources and ephemeral-storage for scale-from-zero specified in MachineClass NodeTemplate
```

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/mcm/mcm_manager.go [hunk 1]
```diff
@@ -689,25 +698,24 @@ func generateInstanceStatus(machine *v1alpha1.Machine) *cloudprovider.InstanceSt
 func validateNodeTemplate(nodeTemplateAttributes *v1alpha1.NodeTemplate) error {
 	var allErrs []error
 
-	capacityAttributes := []v1.ResourceName{"cpu", "gpu", "memory"}
-
-	for _, attribute := range capacityAttributes {
+	for _, attribute := range coreResourceNames {
```
Comment (rishabh-11) [inline cluster-autoscaler/cloudprovider/mcm/mcm_manager.go:701]: FYI - With this GPU is not considered a core resource.

File: cluster-autoscaler/cloudprovider/mcm/mcm_manager.go [hunk 2]
```diff
@@ -689,25 +698,24 @@ func generateInstanceStatus(machine *v1alpha1.Machine) *cloudprovider.InstanceSt
...
-		errMessage := fmt.Errorf("InstanceType, Region and Zone attributes are mandatory")
+		errMessage := fmt.Errorf("InstanceType, Region and Zone attributes are mandatory: %w", ErrInvalidNodeTemplate)
 		klog.Warning(errMessage)
 		allErrs = append(allErrs, errMessage)
 	}
 
 	if allErrs != nil {
-		return fmt.Errorf("%s", allErrs)
+		return errors.Join(allErrs...)
+		//return fmt.Errorf("%s", allErrs)
```
Comment (rishabh-11) [inline cluster-autoscaler/cloudprovider/mcm/mcm_manager.go]:
```suggestion

```

File: cluster-autoscaler/cloudprovider/mcm/mcm_manager.go [hunk 3]
```diff
@@ -1011,3 +1030,16 @@ func isMachineFailedOrTerminating(machine *v1alpha1.Machine) bool {
 	}
 	return false
 }
+
+func filterExtendedResources(allResources v1.ResourceList) (extendedResources v1.ResourceList) {
+	extendedResources = make(v1.ResourceList)
+	for _, n := range knownResourceNames {
```
Comment (rishabh-11) [inline cluster-autoscaler/cloudprovider/mcm/mcm_manager.go]: Aren't we just filtering the `knownResourceNames` in this? What if I have some other extended resource which is not part of this list? I think in general that would be the case

**Review by rishabh-11 — CHANGES_REQUESTED** [review]
Thanks for the PR. some changes requested.

File: cluster-autoscaler/cloudprovider/mcm/mcm_manager.go (hunk 3 above)
Comment (elankath) [reply cluster-autoscaler/cloudprovider/mcm/mcm_manager.go]: We are  the removing the `knownResources` and keeping all other extended resources in this function.  Will add a  doc and unit-test to make this clear.

File: cluster-autoscaler/cloudprovider/mcm/mcm_manager.go (hunk 1 above)
Comment (elankath) [reply cluster-autoscaler/cloudprovider/mcm/mcm_manager.go:701]: Yes, this was stupid oversight. I have `gpuResourceNames` but it makes more correct to have `gpu` inside `coreReourceNames`. To be honest in the real world `Node` we seem to be using only nvidia GPU's labels.

**Comment by elankath** [comment]
Corrected issues. Will test scale from zero for ephemeral storage and custom resources and add manual test-log tomorrow. I am unsure how to code an integration test for this though.

**Comment by elankath** [comment]
Test log for scale from zero with custom resource named `example.com/dongle`


```yaml

apiVersion: v1
kind: Pod
metadata:
  name: testexres1
spec:
  containers:
    - name: example-container
      image: busybox
      command: ["sh", "-c", "echo Using extended resources && sleep 3600"]
      resources:
        limits:
          resource.com/dongle: 2
        requests:
          resource.com/dongle: 2
```

Node Group `b` has this specified:

```yaml

        providerConfig:
          apiVersion: aws.provider.extensions.gardener.cloud/v1alpha1
          kind: WorkerConfig
          nodeTemplate:
            capacity:
              cpu: 8
              ephemeral-storage: 50Gi
              gpu: 0
              hana.hc.sap.com/hcu/cpu: 20
              hana.hc.sap.com/hcu/memory: 10
              memory: 7Gi
              resource.com/dongle: 6

```

Scale from zero triggered.
```

I1119 13:37:51.654540   38814 mcm_manager.go:981] Copying extended resources map[hana.hc.sap.com/hcu/cpu:{{20 0} {<nil>} 20 DecimalSI} hana.hc.sap.com/hcu/memory:{{10 0} {<nil>} 10 DecimalSI} resource.com/dongle:{{6 0} {<nil>} 6 DecimalSI}] to template node.Status.Capacity

 I1119 13:37:41.141731   38814 klogx.go:87] Pod default/testexres1 can be moved to template-node-for-shoot--i062009--abc-b-z1-5762866181449073921-upcoming-0

 Normal   TriggeredScaleUp   37s    cluster-autoscaler  pod triggered scale-up: [{shoot--i062009--abc-b-z1 0->1 (max: 1)}
```

**Review by rishabh-11 — APPROVED** [review]
/lgtm

**Comment by dhague** [comment]
Looking forward to using this feature - thanks!

**Pull request merged**